Add time format with client side date parsing

adamwoodnz opened the pull request:
Closes #413 

This is an alternative approach to the server side implementation in #414 

Pros:
- date description parsing is done at the time of applying the format, rather than on page load on the server
- no filtering of `the_content`

Cons:
- if the date to be parsed by the format needs to be changed, or the post date is changed, the time format needs to be toggled off and on again to re-parse the changed date

UPDATE:
If the time description changes, the format will now be removed so that it can be reapplied and parsed.

adamwoodnz committed:
Add wporg/time block

adamwoodnz committed:
Remove debug

adamwoodnz committed:
Stop parsing content if empty

adamwoodnz committed:
Remove usage of register_block_type as this isn't a block

adamwoodnz committed:
Fix phpcs errors

adamwoodnz committed:
Enqueue frontend script if required rather than rendering inline

adamwoodnz committed:
Change to semantic elements for time on frontend

adamwoodnz committed:
Fix transform_times bail out logic

adamwoodnz committed:
Use Gutenberg non block example approach for registering and enqueuing scripts

adamwoodnz committed:
Remove unused helper

adamwoodnz committed:
Fix bail out logic for no time elements

adamwoodnz committed:
Revert "Use Gutenberg non block example approach for registering and enqueuing scripts"

This reverts commit a5244fde2a1b99fcb512b926dfcb3d78d5461d55.

adamwoodnz committed:
Use Moment to parse the date description and add it to the time element on toggle

adamwoodnz committed:
Replace php dom manipulation with js

adamwoodnz committed:
Update descriptions to specify input format

adamwoodnz committed:
Update docs

adamwoodnz committed:
Skip parsing if format is being deactivated

adamwoodnz committed:
Replace Moment usage with locutus

adamwoodnz committed:
Keep time element within link

tellyworth commented on mu-plugins/blocks/time/index.php:
```diff
@@ -0,0 +1,42 @@
+		'wp_enqueue_scripts',
+		function() use ( $view_script_handle ) {
+			if ( ! is_admin() ) {
+				wp_enqueue_script( $view_script_handle, null, null, filemtime( __DIR__ . '/build/convert_times.js' ), true );
```
It's generally best to use a literal version number instead of `filemtime()` in production, otherwise every page view requires a FS read. We're not exactly consistent about that across the codebase though. Not a blocker.

ryelle replied:
This script is already registered by the `register_block_script_handle` function with the generated dependencies & version string, so you can enqueue it by just using the handle. To also output it in the footer, you can change the `group` — that's all the `$in_footer` [parameter does](https://github.com/WordPress/wordpress-develop/blob/bea4307daa21a3f97d159898b0ac676332e0e7de/src/wp-includes/functions.wp-scripts.php#L173-L184).

```suggestion
				wp_enqueue_script( $view_script_handle );
				wp_script_add_data( $view_script_handle, 'group', 1 );
```

tellyworth commented on mu-plugins/blocks/time/src/convert_times.js:
```diff
@@ -0,0 +1,80 @@
+			const linkElement = document.createElement( 'a' );
+			linkElement.setAttribute(
+				'href',
+				`https://www.timeanddate.com/worldclock/fixedtime.html?iso=${ datetimeISO }`
```
Just noting for future possibilities: should we consider replacing timeanddate.com?

tellyworth commented on mu-plugins/blocks/time/src/index.js:
```diff
@@ -0,0 +1,83 @@
+
+				// Remove the word "at" from the string, if present.
+				// Allows strings like "Monday, April 6 at 19:00 UTC" to work.
+				const dateCleaned = dateDescription.replace( 'at ', '' );
```
(I was curious about why this was here - looks like it's from the original shortcode)

tellyworth reviewed (approved):
Looks fine to me. I've only reviewed the code, not tested.

ryelle commented on mu-plugins/blocks/time/src/convert_times.js:
```diff
@@ -0,0 +1,80 @@
+function convertTimes() {
```
[File names should be separated with a hyphen](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#naming-conventions), so this should be `convert-times.js`. Or you could use `view.js` if this is meant to be the `viewScript`— I don't have a strong opinion about following that convention though.

adamwoodnz replied:
Renamed

ryelle commented on mu-plugins/blocks/time/src/convert_times.js:
```diff
@@ -0,0 +1,80 @@
+		let localTime = '';
+		const datetime = dateElement.getAttribute( 'datetime' );
+		const datetimeISO = dateElement.getAttribute( 'data-iso' );
+		const date = datetime && parseDate( datetime );
```
This was coming up null for me with a valid time: 2023-07-12T19:00:00-04:00

Although, I think that timestamp is wrong, it should either be `2023-07-12T19:00:00-00:00` or `2023-07-12T15:00:00-04:00`

adamwoodnz replied:
Looks like the regex didn't allow negative offsets, which didn't affect me with `+12`. Can you please try again?

adamwoodnz replied:
I've changed my mac to `America/New_York` and with the date string 'Wednesday July 12 2023 at 15:00 EDT' I now see

![Screenshot 2023-07-24 at 9 25 44 PM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/29a76161-f4a0-4117-a50f-7f1ca5f46826)

While with my local timezone I see

![Screenshot 2023-07-25 at 11 52 22 AM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/f92c822b-71b6-4955-a2d2-2582095f32be)

adamwoodnz replied:
This seems consistent with the shortcode version on this draft https://make.wordpress.org/core/?p=106234&preview=1&_ppp=5aa5224312

ryelle reviewed (commented):
The editor side looks great, but it's not working on the frontend, and I think something's off with how the time is converted. I entered "Wednesday July 12 2023 at 15:00 EDT", but the datetime attribute is using the UTC value plus the timezone (`2023-07-12T19:00:00-04:00`).

adamwoodnz committed:
Use wp_script_add_data to output view script in footer

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

adamwoodnz committed:
Rename view script

adamwoodnz commented:
> I entered "Wednesday July 12 2023 at 15:00 EDT", but the datetime attribute is using the UTC value plus the timezone (`2023-07-12T19:00:00-04:00`).

Hmm I hadn't tested with anything other than strings with UTC, as that's what I'd seen in posts. I'll take a look, ideally we could support more than that.

adamwoodnz committed:
Fix regex to allow negative time offsets

adamwoodnz commented on mu-plugins/blocks/time/src/index.js:
```diff
@@ -0,0 +1,86 @@
+
+registerFormatType( name, {
+	title: title,
+	tagName: 'time',
```
FYI I got accessibility team feedback on the old `<abbr title"...">` format and there was general agreement that this `time` element would be better, see [comment on alternative PR](https://github.com/WordPress/wporg-mu-plugins/pull/414#discussion_r1272964500).

adamwoodnz commented:
For some reason this isn't working in my sandbox. The editor script loads and the format appears to be applied (inspecting the element in the editor shows the format), but when I save and load the frontend there is no format applied, and if I reload the editor the format is gone too. It's as if it's being stripped by some tag filter... @ryelle @tellyworth any ideas?

ryelle commented:
> It's as if it's being stripped by some tag filter...

Ahh, I don't see `time` in [the $allowedposttags list](https://github.com/WordPress/wordpress-develop/blob/bea4307daa21a3f97d159898b0ac676332e0e7de/src/wp-includes/kses.php#L68), so I think it _is_ being stripped out. It's not a problem on single sites because those allow unfiltered_html for admins/editors. I think you can add `time` with [the wp_kses_allowed_html filter](https://developer.wordpress.org/reference/functions/wp_kses_allowed_html/).

adamwoodnz committed:
Allow time tags in post content

adamwoodnz committed:
Execute the view script once all resources have loaded

adamwoodnz committed:
Update descriptions

adamwoodnz commented:
> I think you can add `time` with [the wp_kses_allowed_html filter](https://developer.wordpress.org/reference/functions/wp_kses_allowed_html/)

Thanks @ryelle [this seems to work well](https://github.com/WordPress/wporg-mu-plugins/pull/415/commits/3046990443531eeeae98cf7d50e6a06b2968e9d5)

adamwoodnz commented:
> I think you can add `time` with [the wp_kses_allowed_html filter](https://developer.wordpress.org/reference/functions/wp_kses_allowed_html/)

Thanks @ryelle [this seems to work well](https://github.com/WordPress/wporg-mu-plugins/pull/415/commits/3046990443531eeeae98cf7d50e6a06b2968e9d5)

ryelle commented:
Front end view works now. I tried a few different formats, some copied from make/core, and it all worked well  👍🏻 

I know you mentioned you can't reparse an updated date, but is it possible to auto-clear the format when the content changes? I think that's going to be confusing. I tried editing the day in a date and it turned into this:

| Editor | Frontend |
|---|---|
| <img width="673" alt="" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/09d908e1-df95-4329-8ea9-09a0c7781691"> | <img width="729" alt="Screenshot 2023-07-26 at 5 49 23 PM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/e73f7284-d614-4e33-882f-26b818cece72"> |

adamwoodnz commented:
> Front end view works now. I tried a few different formats, some copied from make/core, and it all worked well 👍🏻

🎉 

> is it possible to auto-clear the format when the content changes?

Nice idea, I'll give it a try

adamwoodnz committed:
Style the time formatted strings

adamwoodnz committed:
Add block attribute config

adamwoodnz committed:
Remove the format if the time description is changed

adamwoodnz committed:
Handle multiple times in the selected block

adamwoodnz commented on mu-plugins/blocks/time/postcss/style.pcss:
```diff
@@ -0,0 +1,3 @@
+.wporg-time {
+	text-decoration: underline;
```
I find it's useful to be able to see the content the format has been applied to, and this seems the most appropriate style given that it will end up as a link on the frontend.

adamwoodnz committed:
Use dotted style for the underline

adamwoodnz commented on mu-plugins/blocks/time/postcss/style.pcss:
```diff
@@ -0,0 +1,3 @@
+.wporg-time {
+	text-decoration: underline dotted;
```
I find it's useful to be able to see the content the format has been applied to, and this seems the most appropriate style given that it will end up as a link on the frontend.

StevenDufresne replied:
The block editor content is rendered in an iframe. I don't think this will work.

adamwoodnz replied:
Does it not show for you?

<img src="https://github.com/WordPress/wporg-mu-plugins/assets/1017872/1a5810b4-984a-461f-ae61-ed038703340f" width="500">

adamwoodnz replied:
I see it in the post editor and page editor, but not in the site editor. In fact there are issues with the Gutenberg selectors in the site editor too. I've opened a follow up [issue](https://github.com/WordPress/wporg-mu-plugins/issues/422).

adamwoodnz replied:
I've added partial support in the Site Editor by using inline styles https://github.com/WordPress/wporg-mu-plugins/pull/415/commits/ea1dad6132df73bf94f494d31bf59dee379b5267

I'll do further work in the follow up issue once I have a solution.

adamwoodnz commented:
> is it possible to auto-clear the format when the content changes?

@ryelle I've implemented this now, please try again.

StevenDufresne commented:
I don't know if this is a valid use case, but I got some weird behavior:
1. Successfully create a date,
2. Change `UTC` to `GMT` to simulate a user error
3. Notice you get double dates.

![Screen Capture on 2023-08-07 at 09-32-53](https://github.com/WordPress/wporg-mu-plugins/assets/1657336/7e2cb7e7-37ac-4993-981e-ed4cb832db68)

adamwoodnz commented:
> I got some weird behavior

Thanks for this testing. This is intermittent for me. I did manage to repro once, but after refreshing I can't repro again. I think we can merge and monitor.

adamwoodnz committed:
Add partial support for the Site Editor

Use inline styles in editor and remove on frontend
Handle post date being undefined and parse dates based on current time

adamwoodnz committed:
Update yoast/phpunit-polyfills to satisfy CI

adamwoodnz commented on composer.json:
```diff
@@ -28,7 +28,7 @@
 		"wp-phpunit/wp-phpunit": "~6.0",
 		"wporg/wporg-repo-tools": "dev-trunk",
-		"yoast/phpunit-polyfills": "^1.0"
+		"yoast/phpunit-polyfills": "^1.1"
```
The [PHP unit tests failed](https://github.com/WordPress/wporg-mu-plugins/actions/runs/5932617325) with `Error: Version mismatch detected for the PHPUnit Polyfills. Please ensure that PHPUnit Polyfills 1.1.0 or higher is loaded. Found version: 1.0.5` so I've bumped inline with that

adamwoodnz replied:
Seems like a very new release https://github.com/Yoast/PHPUnit-Polyfills/releases

dd32 replied:
This seems like a safe change to me.  Seems weird that it'd detect the out-of-dateness...

adamwoodnz committed:
Fix editor styles in Safari
